src/bases/createWidgetBase.ts

```diff
@@ -0,0 +1,201 @@
+
+	internalState.historicChildrenMap.forEach((child, key) => {
+		if (!internalState.currentChildrenMap.has(key)) {
+			internalState.historicChildrenMap.delete(key);
+			child.destroy();
+		}
+	});
+	internalState.currentChildrenMap.clear();
+}
+
+const createWidget: WidgetFactory = createStateful
+	.mixin<WidgetMixin, WidgetOptions<WidgetState>>({
+		mixin: {
+			childNodeRenderers: [],
+
+			classes: [],
+
+			getChildrenNodes(this: Widget<WidgetState>): DNode[] {
+				let childrenWrappers: DNode[] = [];
+
+				this.childNodeRenderers.forEach((fn) => {
+					const wrappers = fn.call(this);
+					childrenWrappers = childrenWrappers.concat(wrappers);
+				});
+
+				return childrenWrappers;
+			},
+
+			getNodeAttributes(this: Widget<WidgetState>, overrides?: VNodeProperties): VNodeProperties {
+				const props: VNodeProperties = {};
+
+				this.nodeAttributes.forEach((fn) => {
+					const newProps: VNodeProperties = fn.call(this, assign({}, props));
+					if (newProps) {
+						assign(props, newProps);
+					}
+				});
+
+				if (overrides) {
+					assign(props, overrides);
+				}
+				return props;
+			},
+
+			invalidate(this: Widget<WidgetState>): void {
+				const internalState = widgetInternalStateMap.get(this);
+				if (internalState.dirty) {
+					return;
+				}
+				internalState.dirty = true;
+				this.emit({
+					type: 'invalidate'
+				});
+			},
+
+			get id(this: Widget<WidgetState>): string {
+				const { id } = widgetInternalStateMap.get(this);
+
+				return id || (this.state && this.state.id) || generateID(this);
+			},
+
+			nodeAttributes: [
+				function (this: Widget<WidgetState>): VNodeProperties {
+					const baseIdProp = this.state && this.state.id ? { 'data-widget-id': this.state.id } : {};
+					const { styles = {} } = this.state;
+					const classes: { [index: string]: boolean; } = {};
+
+					const internalState = widgetInternalStateMap.get(this);
+
+					internalState.widgetClasses.forEach((c) => classes[c] = false);
+
+					if (this.state && this.state.classes) {
+						this.state.classes.forEach((c) => classes[c] = true);
+						internalState.widgetClasses =  this.state.classes;
+					}
+
+					return assign(baseIdProp, { key: this, classes, styles });
+				}
+
+			],
+
+			render(this: Widget<WidgetState>): VNode {
+				const internalState = widgetInternalStateMap.get(this);
+				if (internalState.dirty || !internalState.cachedWidget) {
+					const dNode = d(this.tagName, this.getNodeAttributes(), this.getChildrenNodes());
+					const widget = realizeDNode(this, dNode);
+					manageDetachedChildren(this);
+					internalState.cachedWidget = widget;
+					internalState.dirty = false;
+				}
+				return internalState.cachedWidget;
+			},
+
+			tagName: 'div'
+		},
+		initialize(instance: Widget<WidgetState>, options: WidgetOptions<WidgetState> = {}) {
+			const { id, tagName } = options;
+			instance.tagName = tagName || instance.tagName;
+
+			widgetInternalStateMap.set(instance, {
```

Comment:
The question then becomes in what "domain"?  They would only conflict when integrated with something that cares about them being unique (like `app`).  So my feeling is at this level we shouldn't have an opinion about it, but if in something like `app` it would throw/reject situations where duplicates (which `app` already does).